File: 백준/Silver/26549. Holes/Holes.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static class Node {
        private int y;
        private int x;
        
        Node(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }
    
    private static int[] dy = {-1, 0, 1, 0};
    private static int[] dx = {0, 1, 0, -1};
    
    private static int bfs(int i, int j, int r, int c, char[][] board, boolean[][] visited) {
        int space = 0;
        
        Queue<Node> queue = new LinkedList<>();
        
        queue.offer(new Node(i, j));
        visited[i][j] = true;
        
        while (!queue.isEmpty()) {
            Node node = queue.poll();
            space++;
            
            for (int d = 0; d < 4; d++) {
                int y = node.y + dy[d];
                int x = node.x + dx[d];
                
                if (y < 0 || y >= r || x < 0 || x >= c || board[y][x] != '.' || visited[y][x] == true) {
                    continue;
                }
                
                queue.offer(new Node(y, x));
                visited[y][x] = true;
            }
        }
        
        return space;
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        
        int tc = Integer.parseInt(br.readLine());
        
        for (int t = 1; t <= tc; t++) {
            st = new StringTokenizer(br.readLine());
            
            int r = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            
            char[][] board = board_init(r, c, br);
            boolean[][] visited = new boolean[r][c];
            
            int section = 0;
            int space = 0;
            
            // 상자의 섹션과 각 공간의 수 확인
            for (int i = 0; i < r; i++) {
                for (int j = 0; j < c; j++) {
                    if (board[i][j] == '.' && visited[i][j] == false) {
                        section++;
                        space += bfs(i, j, r, c, board, visited);
                    }
                }
            }
            
            StringBuilder sb = new StringBuilder();
            sb.append(section == 1 ? String.format("%d section, ", section) : String.format("%d sections, ", section));
            sb.append(space == 1 ? String.format("%d space", space) : String.format("%d spaces", space));
            
            System.out.println(sb.toString());
        }
    }
    
    private static char[][] board_init(int r, int c, BufferedReader br) throws IOException {
        char[][] board = new char[r][c];
        
        for (int i = 0; i < r; i++) {
            board[i] = br.readLine().toCharArray();
        }
        
        return board;
    }
}

File: 백준/Silver/26549. Holes/Holes.py
from collections import deque

def bfs(i, j, r, c, board, visited):
    space = 0
    queue = deque()

    queue.append((i, j))
    visited[i][j] = True

    while queue:
        ny, nx = queue.popleft()
        space += 1

        for dy, dx in [(-1, 0), (0, 1), (1, 0), (0, -1)]:
            y = ny + dy
            x = nx + dx

            if 0 <= y < r and 0 <= x < c and board[y][x] == '.' and visited[y][x] == False:
                queue.append((y, x))
                visited[y][x] = True

    return space

for _ in range(int(input())):
    r, c = map(int, input().split())

    board = [[j for j in input()] for _ in range(r)]
    visited = [[False] * c for _ in range(r)]
    
    section = 0
    space = 0

    for i in range(r):
        if board[i][0] == '.' and visited[i][0] == False:
            bfs(i, j, r, c, board, visited)
        if board[i][c - 1] == '.' and visited[i][c - 1] == False:
            bfs(i, j, r, c, board, visited)
    
    for j in range(c):
        if board[0][j] == '.' and visited[0][j] == False:
            bfs(i, j, r, c, board, visited)
        if board[r - 1][j] == '.' and visited[r - 1][j] == False:
            bfs(i, j, r, c, board, visited)

    # 상자 섹션과 각 공간의 수 확인
    for i in range(r):
        for j in range(c):
            if board[i][j] == '.' and visited[i][j] == False:
                section += 1
                space += bfs(i, j, r, c, board, visited)

    result = ''
    result += f'{section} sections, ' if section != 1 else f'{section} section, '
    result += f'{space} spaces' if space != 1 else f'{space} space'

    print(result)
